File: implementations/python/src/lumen/cadencia.py
"""
Cadencia Bridge client — spawns the Rust sidecar for high-performance file indexing.

The sidecar reads files from disk, compresses them with LUMEN, and returns
binary frames. The Python client communicates with the sidecar via a
JSON line-delimited protocol over stdin/stdout.

Ported from TypeScript ``src/cadencia.ts``.
"""

from __future__ import annotations

import asyncio
import json
import logging
import subprocess
import sys
from dataclasses import dataclass, field
from typing import Any

logger = logging.getLogger(__name__)


# ═══ Types ════════════════════════════════════════════════════════════════════


@dataclass
class BridgeCommand:
    """A command sent to the Cadencia sidecar."""

    cmd: str
    files: list[str] = field(default_factory=list)


@dataclass
class BridgeResponse:
    """A response from the Cadencia sidecar."""

    status: str


@dataclass
class BridgeIndexResponse(BridgeResponse):
    """Response from an index command."""

    files: int = 0
    total_bytes: int = 0
    wire_bytes: int = 0
    encode_us: int = 0


@dataclass
class BridgeOptions:
    """Options for the Cadencia bridge."""

    binary_path: str
    """Path to the compiled ``cadencia-bridge`` binary."""
    cwd: str | None = None
    """Working directory for the sidecar process."""


# ═══ CadenciaBridge ═══════════════════════════════════════════════════════════


class CadenciaBridge:
    """Manages the lifecycle of the Cadencia sidecar process.

    Usage::

        bridge = CadenciaBridge(BridgeOptions(binary_path="./cadencia-bridge"))
        await bridge.start()
        result = await bridge.index(["src/main.ts", "src/utils.ts"])
        await bridge.stop()
    """

    def __init__(self, options: BridgeOptions) -> None:
        self._opts = options
        self._process: subprocess.Popen[bytes] | None = None
        self._reader_task: asyncio.Task[None] | None = None
        self._response_queue: asyncio.Queue[dict[str, Any]] = asyncio.Queue()
        self._protocol_version: str = ""

    async def start(self) -> dict[str, Any]:
        """Spawn the sidecar and perform handshake."""
        kwargs: dict[str, Any] = {
            "args": [self._opts.binary_path],
            "stdin": subprocess.PIPE,
            "stdout": subprocess.PIPE,
            "stderr": subprocess.PIPE,
            "cwd": self._opts.cwd,
        }

        self._process = subprocess.Popen(**kwargs)
        self._reader_task = asyncio.create_task(self._read_responses())

        # Ping handshake
        result = await self._send_command("ping")
        self._protocol_version = result.get("protocol", "unknown")
        return result

    async def index(self, files: list[str]) -> BridgeIndexResponse:
        """Index a list of files through the sidecar."""
        result = await self._send_command("index", files=files)
        return BridgeIndexResponse(
            status=result.get("status", "error"),
            files=result.get("files", 0),
            total_bytes=result.get("total_bytes", 0),
            wire_bytes=result.get("wire_bytes", 0),
            encode_us=result.get("encode_us", 0),
        )

    async def stop(self) -> dict[str, Any]:
        """Graceful shutdown."""
        result = await self._send_command("stop")
        if self._reader_task:
            self._reader_task.cancel()
            try:
                await self._reader_task
            except asyncio.CancelledError:
                pass
        if self._process:
            try:
                self._process.stdin.close()
            except Exception:
                pass
            try:
                self._process.wait(timeout=3)
            except subprocess.TimeoutExpired:
                self._process.kill()
                self._process.wait()
            self._process = None
        return result

    async def _send_command(self, cmd: str, files: list[str] | None = None) -> dict[str, Any]:
        """Send a command and wait for the response."""
        assert self._process and self._process.stdin

        command: dict[str, Any] = {"cmd": cmd}
        if files:
            command["files"] = files

        line = json.dumps(command) + "\n"
        self._process.stdin.write(line.encode("utf-8"))
        self._process.stdin.flush()

        try:
            return await asyncio.wait_for(self._response_queue.get(), timeout=30)
        except asyncio.TimeoutError:
            raise TimeoutError(f"Cadencia bridge did not respond to '{cmd}' within 30s")

    async def _read_responses(self) -> None:
        """Read line-delimited JSON responses from stdout."""
        assert self._process and self._process.stdout

        loop = asyncio.get_running_loop()
        buf = b""

        while True:
            try:
                chunk = await loop.run_in_executor(
                    None, self._process.stdout.read, 4096
                )
            except Exception:
                return

            if not chunk:
                return

            buf += chunk
            while b"\n" in buf:
                line, buf = buf.split(b"\n", 1)
                line = line.strip()
                if not line:
                    continue
                try:
                    msg = json.loads(line)
                    await self._response_queue.put(msg)
                except json.JSONDecodeError:
                    logger.warning("Cadencia bridge: malformed JSON line: %s", line)

File: implementations/typescript/src/cadencia.ts
/**
 * Cadencia Bridge Client — spawns the Rust `cadencia-bridge` sidecar
 * and communicates via line-delimited JSON on stdin/stdout.
 *
 * This is how the VS Code extension (TypeScript) offloads heavy
 * I/O and LUMEN encoding to a native Rust process.
 *
 * Protocol (see cadencia-bridge.rs for the Rust side):
 * ```
 * TS → Rust:  {"cmd":"index","files":["path1","path2",...]}
 * Rust → TS:  {"status":"ok","files":1234,"total_bytes":...}
 * ```
 */

import { spawn, ChildProcess } from "child_process";
import { createInterface } from "readline";

// ── Types ────────────────────────────────────────────────────────────────────

export interface BridgeCommand {
  cmd: "ping" | "index" | "stop";
  files?: string[];
  id?: number;
}

export interface BridgeResponse {
  status: "ok" | "error";
  version?: string;
  protocol?: string;
  files?: number;
  total_bytes?: number;
  wire_bytes?: number;
  encode_us?: number;
  elapsed_us?: number;
  error?: string;
  reason?: string;
  id?: number;
}

export interface BridgeOptions {
  /** Path to the cadencia-bridge binary. */
  binaryPath: string;
  /** Working directory for file resolution. */
  cwd?: string;
  /** Called when the bridge process exits unexpectedly. */
  onCrash?: (code: number | null) => void;
}

// ── Client ───────────────────────────────────────────────────────────────────

export class CadenciaBridge {
  private proc: ChildProcess | null = null;
  private pending: Map<
    number,
    { resolve: (r: BridgeResponse) => void; reject: (e: Error) => void }
  > = new Map();
  private seq = 0;

  constructor(private options: BridgeOptions) {}

  /** Start the sidecar process. Returns the ready response. */
  async start(): Promise<BridgeResponse> {
    this.proc = spawn(this.options.binaryPath, [], {
      cwd: this.options.cwd,
      stdio: ["pipe", "pipe", "pipe"],
    });

    const rl = createInterface({ input: this.proc.stdout! });
    rl.on("line", (line: string) => {
      try {
        const msg: BridgeResponse = JSON.parse(line);
        // Match response to its pending request by sequence id
        if (msg.id !== undefined && this.pending.has(msg.id)) {
          const { resolve } = this.pending.get(msg.id)!;
          this.pending.delete(msg.id);
          resolve(msg);
        } else {
          // Unsolicited response — log and drop
          console.warn(
            `CadenciaBridge: unsolicited response (id=${msg.id ?? "none"})`
          );
        }
      } catch {
        // ignore malformed lines
      }
    });

    this.proc.on("exit", (code) => {
      this.options.onCrash?.(code);
      // Reject all pending
      for (const [, { reject }] of this.pending) {
        reject(new Error(`Bridge exited with code ${code}`));
      }
      this.pending.clear();
    });

    // The bridge sends a ready signal immediately on start
    return this.send({ cmd: "ping" });
  }

  /** Index a list of files. Returns timing and wire stats. */
  async index(files: string[]): Promise<BridgeResponse> {
    return this.send({ cmd: "index", files });
  }

  /** Stop the sidecar gracefully. */
  async stop(): Promise<void> {
    try {
      await this.send({ cmd: "stop" });
    } finally {
      this.proc?.kill();
    }
  }

  private send(cmd: BridgeCommand): Promise<BridgeResponse> {
    return new Promise((resolve, reject) => {
      if (!this.proc || !this.proc.stdin) {
        reject(new Error("Bridge not started"));
        return;
      }
      const seq = ++this.seq;
      this.pending.set(seq, { resolve, reject });
      const wire = { ...cmd, id: seq };
      this.proc.stdin.write(JSON.stringify(wire) + "\n");

      // Timeout after 30s
      setTimeout(() => {
        if (this.pending.has(seq)) {
          this.pending.delete(seq);
          reject(new Error("Bridge request timed out"));
        }
      }, 30_000);
    });
  }
}
